### DISCUSSION TITLE
Macro defining syntax proposal

### DISCUSSION THREAD

**Discussion by tcallred** [racket/rhombus#213] [Ideas]
In response to the recent discussion on macros on Rhombus, I would like to share my own ideas on how we can make macro definitions more ergonomic and user friendly. The main idea is to remove incidental details so that the programmer can focus on essential parts of the macro definition. 

Today, defining a macro looks something like this (from the docs):
```
fun
| factorial(0): 1
| factorial(n): n*factorial(n-1)

expr.macro '($a *! $tail ......):
  values('(factorial($a)), tail)

10*!
```
Defining a macro requires using `expr.macro` with a template denoted by `'( ... )`, and a block that returns `values` where the first argument is a template with the resulting expanded code and the second argument is matched syntax to be handled by something else. (You can also define a macro using `def` without the need for `values`). 

In this post, I would like to propose a few changes to the syntax for the sake of discussion. To do that, I will use a toy-example macro that imitates python's list comprehensions. 

(In today's macro system)
```
expr.macro '($exp for $a ~in $lst):
  values('(base.map(fun($a): $exp , $lst)), '())

(x*x) for x ~in [1, 2, 3] // -> [1, 4, 9]
```

## Support for single return
When writing a macro that consumes all of its matched input and would return `values( ..., '())`, it would be nice to be able to simply return the template without the need to call values. 
```
expr.macro '($exp for $a ~in $lst):
  '(base.map(fun($a): $exp , $lst))
```
This would implicitly produce `values('(base.map(fun($a): $exp , $lst)), '())`. This would make it similar to how `def` creates macros when given a template and a body with a resulting template.  It also just makes it a little bit cleaner. 

## Templates created with back-ticks or single-quotes

Instead of using single-quote as a prefix operator paired with parentheses `'(...)` to create a template, we could use back-ticks or single-quotes as a new type of parentheses. This would require an extension to shrubbery but could make templates cleaner by removing extra parentheses. 

My first idea involved using back-ticks, but after some discussion with @mflatt, it seemed like it would be better to use single quotes, specifically to avoid clashing with markdown. Below, I've included examples using both for the sake of discussion. 

Using back-ticks:
```
expr.macro `$exp for $a ~in $lst`:
  `base.map(fun($a): $exp , $lst)`
```
Pros:
- Removing extra parentheses cleans it up by removing distraction from the content of the template
- Back-tics are typically associated with "code" and sends the message of something "different" from a string
- Relates to JavaScript's interpolated strings, helping people make the mental connection between macro templates and string interpolation. 

Cons:
- Would require a major extension to shrubbery. 
- Would conflict with the meaning of back-ticks in markdown documentation

Using single-quote
```
expr.macro '$exp for $a ~in $lst':
  'base.map(fun($a): $exp , $lst)'
```
Pros:
- Maintains the cleanliness just like back-tick
- Does not conflict with markdown syntax
- Would be relatively easy to extend shrubbery for considering that today single-quote is only read as an operator and could be changed act like parens 
- Relates to Racket's `'()` forms

Cons:
- Would still need to change shrubbery
- Single quotes might be confusing for those used to using them for strings or chars
- Harder to distinguish from double-quote

## Other examples with proposed ideas
```
expr.macro '$a *! $tail ... ':
  values( 'factorial($a)', tail )
```
```
expr.macro 'prefix_plus $a $b $c ...':
  values( a, '+ $b $c ...' )
```
```
expr.macro 'also_prefix_plus $e ...':
  match e
  | '$a $b $c ...':
      values( a, '+ $b $c ...' )
  | ~else:
      '"this is terrible error reporting"'
```

**Comment by sorawee** [comment]
`expr.macro` is for low-level macro. You can use [`expr.rule`](https://plt.cs.northwestern.edu/pkg-build/doc/rhombus/Expression_Macros.html#%28def._%28%28lib._rhombus%2Fmacro..rkt%29._expr._%28rule._~23f%29%29%29) which already "supports for single return"

```
#lang rhombus

import:
  rhombus/macro open

fun
| factorial(0): 1
| factorial(n): n*factorial(n-1)

expr.rule '($a *!):
  '(factorial($a))

10*!
```

**Comment by mflatt** [comment]
I like the idea of having `'` as a kind of parenthesis and using it for syntax quotes. That change would make things less noisy by removing the need for some parentheses, it helps avoids having some character combinations inconveniently treated as a single operator (such as `'$` versus `'` followed by `$`), and it removes the need for something like `syntax_term` to represent a term without parentheses. Also, balanced quotes are probably look more familiar to many. Although it would mean a new reserved character in shrubbery notation, it looks like a straightforward change, and having four kinds of parentheses instead of three kinds doesn't seem bad.

**Reply by sorawee** [reply]
Currently, we can do 
```
'('(f(1)))
```

If we use `'` as a parenthesis, how would we write this?

**Reply by mflatt** [reply]
One possibility is
```
'( 'f(1)' )'
```
You could read this as `( '` being an opener for a nested quote that's closed by `' )`.

**Comment by samdphillips** [comment]
My concern with this proposal is that using a single token to represent the opening and closing of a syntax quote will make writing a macro that expands to something containing a syntax quote difficult to write/read.

**Reply by mflatt** [reply]
I think the difference compared to now would mostly be adding a `'` at the end. That's because you almost always have to write  `(` after `'` currently (for all but the simplest quoted terms). That `(` with its closing `)` would prevent a nested `'` from matching outside the parentheses.

**Comment by gus-massa** [comment]
It's probably 50 year too late to change this, but I hate that most (all?) programming languages use the same character to start and end an string `"Hello"`, and LaTeX use the same symbol to start and end math mode `$x^2$`.

(Preemptive reply: Yes, but nobody uses `\(x^2\)`. Moreover, `$$x^2$$` is even worse.)

**Reply by benknoble** [reply]
Not sure why this is relevant to the discussion being had here, but FWIW I have `$` bound to insert `\(` and `\)` as appropriate (in vim), so some people do in fact use it.

**Reply by rocketnia** [reply]
This has been a pet peeve of mine too, as one of a few long-standing things I think could use fine-tuning in string syntaxes to let escape sequences be avoided in common cases. The string syntaxes I've pursued look more like at-expressions.

That said, I think the `'...'` syntax quotes look surprisingly conventional, even cozy. I like this change from a conventional syntax perspective, especially knowing that at-expressions and the Racket reader are right there as alternatives for more niche kinds of quoting concerns.

**Comment by mflatt** [comment]
I've pushed the change to use `'`...`'` for quoting syntax. The "demo.rhm" diff at the top of f21679842c56b0841b156dde6b389e0366f15532 (you'll probably have to click to load the diff) illustrates the improvement.

As one example, the macro
```
defn.macro '(def_five $id):
  '(:
      def $id: 5
  )
```
becomes
```
defn.macro 'def_five $id':
  'def $id: 5'
```


Initially, I expected to push this change in a branch, so we could try out it. But it ended up being an involved change, and part of the reason things get better is that the change enabled and forced a generalization of syntax objects to gracefully handle single-term, single-group, and multi-group syntax objects. So, whatever syntax we settle on, I'm pretty sure the path runs through this version of the implementation.

The generalization to different kinds of syntax objects is why `def_five` above gets so much better: the `def $id 5` part doesn't have to be packaged in a fake block. Multiple result groups work the way you'd expect:

```
defn.macro 'def_five_and_six $id5 $id6':
  'def $id5: 5
   def $id6: 6'
```

**Reply by mflatt** [reply]
While I'm breaking things, I also changed `$:` to just `::` for syntax-class annotations.

**Reply by Lazerbeak12345** [reply]
How would a macro-generating-macro look?

```txt
defn.macro 'def_a_and_b $ida $idb':
  'defn.macro \'def_$ida_and_$idb \$id5 \$id6\':
    \'def \$id5: 5
     def \$id6: 6\''
```

**Reply by mflatt** [reply]
There's no "identifier interpolation" as you seem to have in mind with `def_$ida_and_$idb`, but here's how the nested quotes using `'«` ... `»'` and escaped escapes using `$('$')` work out:

```
defn.macro 'def_a_and_b $def_form':
  '«defn.macro '$def_form $('$')id5 $('$')id6':
     'def $('$')id5: 5
      def $('$')id6: 6'»'

def_a_and_b def_five_and_six
def_five_and_six five six
five
six
```

**Reply by sorawee** [reply]
That looks kinda non-ergonomic to write.

**Reply by Lazerbeak12345** [reply]
Wouldn't identifier interpolation be a desirable feature? I know of a few racket macros that rely on similar behavior in racket.

If identifier iterop. were to be considered, it'd have to be done in a way that prevents collision with other potentially preexisting identifiers - something that might just be handled by the behavior of the host language's `define` equivalent (keeping in mind that typed/rombus will very likely be a thing in the future). This, I believe, is what racket currently does.

On the other hand, I do like that there is potential to use a specific identifier that the user wishes. Perhaps this detail merits further research.

**Reply by mflatt** [reply]
> That looks kinda non-ergonomic to write.

I forgot to mention that this iteration tries out yet another change, which is making nested `''` not increase the number of `$`s needed to escape. If we change quotes back to increasing the nesting depth, back to being like `quasiquote`, then the `$('$')` go back to being `$`. Each choice (increasing escape depth or not) makes different things more difficult, but both seem workable. Getting rid of the `«»`, meanwhile, probably requires using more characters to have a distinct opener and closer.

> Wouldn't identifier interpolation be a desirable feature?

Yes, that sound's useful, but it's not there so far.

**Reply by Lazerbeak12345** [reply]
One could `(define $ '$')` and perhaps that would turn your latest example into this:

```
defn.macro 'def_a_and_b $def_form':
  '«defn.macro '$def_form $($)id5 $($)id6':
     'def $($)id5: 5
      def $($)id6: 6'»'

def_a_and_b def_five_and_six
def_five_and_six five six
five
six
```

It seems a touch less verbose, but in this context I actually think this might be good too:

```
defn.macro 'def_a_and_b $def_form':
  '«defn.macro '$def_form $$id5 $$id6':
     'def $$id5: 5
      def $$id6: 6'»'

def_a_and_b def_five_and_six
def_five_and_six five six
five
six
```

**Reply by Lazerbeak12345** [reply]
My concern, however is that, however elegant `'«` `»'`  is,

1. It's more or less equal to parens, but a touch better by having closing annotation unique to templates.
   - A macro-generating-macro-generating-macro might be hard to read as it would end with `»'»'`. seems verbose somehow.
2. `»'` is hard to type. Would `>>'` serve to be equivalent? (as `lambda` and `λ` are in `lang/racket`)

**Reply by plane** [reply]
I like it, but `'«` and `»'` are a little hard on my poor brain, particularly when they're right next to a `'`, like in this example.

I'm just going to share an idea, even though I don't think many people will like it.  This is what it looks like with `｢` and `｣`, which look very different from the other quote pairs, so they're easier to spot:

```
defn.macro 'def_a_and_b $def_form':
  ｢defn.macro '$def_form $('$')id5 $('$')id6':
     'def $('$')id5: 5
      def $('$')id6: 6'｣
```

**Reply by sorawee** [reply]
We can use "｢ and ｣" everywhere too.

```
defn.macro ｢def_a_and_b $def_form｣:
  ｢defn.macro ｢$def_form $(｢$｣)id5 $(｢$｣)id6｣:
     ｢def $(｢$｣)id5: 5
      def $(｢$｣)id6: 6｣｣
```

And with `$$` as a shorthand for `$(｢$｣)`:

```
defn.macro ｢def_a_and_b $def_form｣:
  ｢defn.macro ｢$def_form $$id5 $$id6｣:
     ｢def $$id5: 5
      def $$id6: 6｣｣
```

But actually, if we want to stick with ASCII (and I think we should), this could work too:

```
defn.macro {def_a_and_b $def_form}:
  {defn.macro {$def_form $$id5 $$id6}:
     {def $$id5: 5
      def $$id6: 6}}
```

Of course, this clashes with the dictionary / set literal, but to be honest, I'm willing to pay that price. Sets are not frequently used anyway. And I'd prefer a more extensible approach like this:

```
let x: set[1, 2, 3]
let y: dict[1: 2, 3: 4, 5: 6]
let z: list[1, 2, 3]
let w: your_container["a", "b"]
```

over throwing away many useful delimiters for diminishing return.

**Reply by Lazerbeak12345** [reply]
Isn't "｢ and ｣" really just an alias for a function name anyway? `syntax`?

I like what you're going for, but one thing is that I'd imagine that people would find more regular use of sets then they would macros - or does this syntax apply to all templates?

IE @mflatt are you saying `defn.macro` should take a `list` object? Doesn't that mean we loose debugging context (line and column info for where tokens or blocks were originally)?

**Reply by mflatt** [reply]
> you saying defn.macro should take a list object?

No,  it's all syntax objects.

**Reply by mflatt** [reply]
As of last week's meeting, my impression was that having quotes increase nesting depth would make macro-defining macros work. But I was missing that `def_form` would not be escaped in this kind of example:

```
defn.macro 'def_a_and_b $def_form':
  '«defn.macro '$def_form $id5 $id6':
     'def $id5: 5
      def $id6: 6'»'
```

So, here's another idea: square brackets around a `$` argument could indicate extra escape levels:

```
defn.macro 'def_a_and_b $def_form':
  '«defn.macro '$[def_form] $id5 $id6':
     'def $id5: 5
      def $id6: 6'»'
```

This choice of notation doesn't collide with existing things, currently. Square brackets would normally create a list, and `$` doesn't coerce a list  automatically to a syntax object (partly because it's not clear what it could be converted to, unlike the S-expression case).

This approach also works well for `scribble/rhombus`, where having quotes increase nesting depth created enough trouble to inspire my attempt to do without. Adding `[]` in the two places where I need an extra escape level is straightforward.

The approach doesn't help at all for avoiding `'«` ... `»'`, as far as I can tell.

**Reply by rocketnia** [reply]
I think the `$[def_form]` notation looks fine, and macro-generating macros are rare enough that something like this might be more than enough for real programs, but I think it might not be ideal. (I'm going to bring up some more unsual design thoughts I've been having for Cene.)

Generally, it would be easy to write DSLs in most languages if code of the form `lambda(list(local("x")), call(global("+"), list(local("x"), local("x"))))` were tolerable to maintain. I think one of the main appeals of quoted s-expressions for DSL purposes is the way they allow variables to be written `x` rather than having all these delimiters. I think this is also why lots of programmers who otherwise prefer s-expressions still find `,x` more readable than `(unquote x)`.

If the language were to have no particular support, programmers could probably get by with something like `$(quote_syntax($def_form))`, essentially replicating the `,',def_form` technique for escaping from multiple levels of quasiquotation. But unlike `,',def_form`, something like `$(quote_syntax($def_form))` is still more like `(unquote x)` than `,x` in terms of clutter.

I think `$[def_form]` is better but still has some of that clutter remaining in the form of brackets.

If people consider it reasonable, I think it might be better to put all the clutter on the left side, such as `$!>!def_form`, `$.>.def_form`, `$.def_form`, `$-def_form`, or `${>}def_form`. That allows the overall snippet to have a jagged silhouette, more like the silhouette a simple identifier would have had, making it potentially less distracting from the code the programmer is really trying to envision when they write it.

I listed a bunch of possibilities there. In the ones that use `!` and `{`, I'm picking relatively rough-looking separators so they don't visually break up the "identifier" into pieces as crisply as a `(` or `|` would. In the ones that use `.`, I'm attempting to appeal to the existing conventional use of `.` as a namespace qualification operator for identifiers.

And the point of `>` is to allow us to write sequences like `$!>>>>!def_form` when we want to escape through multiple stages of quotation at a time. (In `$.def_form` and `$-def_form`, I'm being a bit bolder and imagining the use of something a little terser like `$....def_form` or `$----def_form`.)

I think each `>` should have a corresponding `<` annotating the pair of quotation brackets that it escapes from. Instead of writing the opening bracket as `'`, we'd write it as `$!<!'` or `'!<!$`, or maybe the terser `'-$` or `$-'` if it's matching up with something like `$-def_form`. (I'm using `$` as part of the notation for annotated quotation brackets here so that it's possible for *escaped brackets* to resemble both unescaped brackets and other escape sequences.)

To keep this from being endlessly confusing, I think unannotated uses like `'` and `$foo` should match up, and annotated uses like `$!<!'` and `$!>!foo` should match up, but they shouldn't pass through each other or interact in any way. Basically, they should be like `(`..`)` and `[`..`]`, where a mismatched `(`..`]` is always treated as a mistake. And although `$!>>>>!foo` should interact with multiple `$!<!'` brackets, the `$!<!'` brackets it interacts with should have no unannotated `'` brackets in between.

Overall, since majority of macros aren't macro-generating macros, this lets most code completely ignore the `<` annotations. The code that actually needs to manage multiple stages of quotation must start explicitly annotating every stage rather than using a potentially confusing mix of explicit and implicit interactions.

Actually, I think there's a deeper semantic purpose to keeping things explicit here: For some kinds of quotation, escaping from multiple layers of quotation might not have semantics quite as simple as `$(quote_syntax($def_form))`. It's possible some stages might use some slightly different kind of re-quotation there, like `(quote-syntax def_form #:local)` or `(syntax/loc stx def_form)`, and some stages might perform sanitization or ascribe types or contracts to the things passing through. Making multiple-stage leaps more explicit using `<` and `>` should make it more obvious where in the code these other subtle choices can be written down and reviewed.

To make this all a bit more specific and tangible, this is what I'm imagining:

```
defn.macro 'def_a_and_b $def_form':
  $!<!｢
    defn.macro '$!>!def_form $id5 $id6':
      '
        def $id5: 5
        def $id6: 6
      '
  ｣
```

Besides the `$!>!` stuff, this example also incorporates some shallower stylistic tweaks:

* I've used `｢` and `｣` here for one of the pairs of quotation brackets. I think picking another Unicode character like this might be better than doubling up on the meaning of `«`. (Maybe Rhombus is already using `!` too, in which case another character might be better there.)

* I've indented by a whole tab stop at a time. I often work with text editors like Notepad++ and Visual Studio Code that make it a pain to indent by variable numbers of spaces, and in terms of readability, I've always had to do a double take when I see two adjacent lines that differ by only one space of indentation, since that looks so close to being a mistake.

* I've used an indentation policy where if a pair of matching brackets spans multiple lines, then the lines it begins and ends on have the same amount of indentation. I think this tends to reflect how codebases in languages like C treat their curly brackets (though not always the way they treat their function call parentheses). In this case, besides looking more conventional, it avoids an asymmetrical phenomenon that would otherwise be caused by my whole-tab-stop-oriented indentation style, where opening brackets end up leading a much more wallflowery life than their corresponding closing brackets:
  ```
  defn.macro '$!>!def_form $id5 $id6':
    '
      def $id5: 5
      def $id6: 6'
  ```

**Reply by mflatt** [reply]
In trying out documentation updates for `$[]`, I discovered that I had defined `...` to also be sensitive to nesting depth — which seems natural and consistent with `$`, but runs into the same trouble as `$` escape levels and doesn't have a natural place to add `[]`. In Racket, meanwhile, `...` is not sensitive to nesting depth; there's a totally different and special (and not popular) mechanism for quoting `...` in templates. So, after trying out some alternatives, I now think that `$[]` may be trying too hard, while  `$('$')` plus `$('...')` is not so bad and is at least simple and consistent.

@rocketnia's idea of a family of quoting an unquoting forms, like `@|<<{` ... `}>>|` with escaping by `|>>@`, has worked well for reader-level quoting for at-notation. But for this to be used for patterns and templates, I think it would have to be lifted to a notion of parenthesis families and accompanying operators at the level of shrubbery "output" (i.e., the representation of a shrubbery term). that seems like a lot of extra complexity, too, and there would still be the `...` problem.

**Reply by Lazerbeak12345** [reply]
I'm actually warming up to `$('$')` syntax. reasons are, primarily that 

- Macros inside macros are _very_ rare.
- It's simpler to parse
- It is making use of normal escaping, in a sense.

**Reply by rocketnia** [reply]
> But for this to be used for patterns and templates, I think it would have to be lifted to a notion of parenthesis families and accompanying operators at the level of shrubbery "output" (i.e., the representation of a shrubbery term). that seems like a lot of extra complexity, too[...]

Yeah, that's where the complexity would go, and it would be a bunch of complexity for a niche purpose.

> the `...` problem

Hmm, that's also an escape sequence. Some options that you've probably already considered:

* `$[foo] [...]`
* `[$[foo]] ...`
* `$[foo] ...[]`
* `$[foo] []...`
* `📜$foo 📜...' (just picking some rare character to serve as a third escape sequence marker that interacts with the others)
* `$$foo $...` (based on @gus-massa's `$$foo` suggestion [edit: which I just noticed @Lazerbeak12345 suggested above])
* Redesign it to be written `$...`, and go with:
  * `$[foo] $[...]`
  * `$$foo $$...`

> So, after trying out some alternatives, I now think that `$[]` may be trying too hard, while `$('$')` plus `$('...')` is not so bad and is at least simple and consistent.

Would these not give errors like "expected expression after `$` in `'$'`" and "expected template term before ellipses in `'...'`"? Not that there's anything wrong with having a special case for something like this--all the other options seem to be special cases in their own ways--but I wonder if this has been taken into account.

**Reply by mflatt** [reply]
> Would these not give errors like "expected expression after `$` in `'$'`" and "expected template term before ellipses in `'...'`"?

No. As you say, a special case has been carved out for `$` and `...` by themselves in a template:

```
> '$'
'$'
> '...'
'...'
> '... x'
string:1:1: ...: misplaced repetition
  in: ...
 [,bt for context]
```

The following current behavior is is an accidental, though, and probably should be changed.

```
> 'x $'
'x $'
```

**Comment by gus-massa** [comment]
I like the general idea of rhombus of using indentation to define blocks, but I think that blocks of code in macros are weird enough to go back to matched parenthesis. It's slightly similar to strings, that really need a starting and an ending border.

I insist in using a different delimiter for each border, so nesting is easier. And my idea is to use `$$`to escape two levels instead of one, with the obvious generalization to `$$$`...

I'm not sure if this makes sense, but my proposal is 

    defn.macro «def_a_and_b $def_form»:
      «defn.macro «$$def_form $id5 $id6»:
         «def $id5: 5
          def $id6: 6»»
